packages/core/contracts/financial-templates/perpetual-multiparty/PositionManager.sol

```diff
@@ -496,39 +474,48 @@ contract PositionManager is FeePayer, AdministrateeInterface {
         tokenCurrency.burn(numTokens.rawValue);
     }
 
+    function repay(FixedPoint.Unsigned memory numTokens)
+        public
+        notEmergencyShutdown()
+        noPendingWithdrawal(msg.sender)
+        fees()
+        nonReentrant()
+        returns (FixedPoint.Unsigned memory amountRepaired)
+    {
+        // TODO: implementation
+    }
+
     /**
-     * @notice After a contract has passed expiry all token holders can redeem their tokens for underlying at the
-     * prevailing price defined by the DVM from the `expire` function.
+     * @notice If the contract is emergency shutdown then all token holders and sponsors can redeem their tokens or
+     * remaining collateral for underlying at the prevailing price defined by a DVM vote.
      * @dev This burns all tokens from the caller of `tokenCurrency` and sends back the proportional amount of
      * `collateralCurrency`. Might not redeem the full proportional amount of collateral in order to account for
      * precision loss. This contract must be approved to spend `tokenCurrency` at least up to the caller's full balance.
      * @return amountWithdrawn The actual amount of collateral withdrawn.
      */
-    function settleExpired()
+    function settleEmergencyShutdown()
         external
-        onlyPostExpiration()
+        onlyEmergencyShutdown()
         fees()
         nonReentrant()
         returns (FixedPoint.Unsigned memory amountWithdrawn)
     {
-        // If the contract state is open and onlyPostExpiration passed then `expire()` has not yet been called.
-        require(contractState != ContractState.Open, "Unexpired position");
-
-        // Get the current settlement price and store it. If it is not resolved will revert.
-        if (contractState != ContractState.ExpiredPriceReceived) {
-            expiryPrice = _getOraclePrice(expirationTimestamp);
-            contractState = ContractState.ExpiredPriceReceived;
+        // Set the emergency shutdown price as resolved from the DVM. If DVM has not resolve will revert.
+        if (emergencyShutdownPrice.isEqual(FixedPoint.fromUnscaledUint(0))) {
```

Comment:
So if an identifier actually resolves to 0, then this function will always call `_getOracleEmergencyShutdownPrice()`, which is not a problem, but an inefficiency.

Reply:
ye, good point. I dont think it's a huge issue though as we still support 0 prices (for binary outcomes for example).

Reply:
yeah, maybe just add a test for a price that resolves to 0 and make sure it doesn't revert?

Reply:
sure. I've added a test to check we can support 0 price resolution. sponsors in this case get back all their collateral, irrespective of token debt and token holders get nothing as their tokens are worth 0.

Reply:
A+ financial product

Reply:
binary outcome on a perp!